lib/tree.rb

```diff
@@ -16,49 +16,153 @@ def initialize
+
+  # Time Complexity: O(log n) where n is number of nodes in tree
+  # Space Complexity: O(log n) where n is number of nodes in tree
+  def find_helper(current_node, search)
+    return nil if current_node.nil?
+    
+    if current_node.key == search
+      return current_node
+    elsif current_node.key > search
+      find_helper(current_node.left, search)
+    else
+      find_helper(current_node.right, search)
+    end
+  end
+  
   def find(key)
-    raise NotImplementedError
+    found = find_helper(@root, key)
+    return found.value if found
+  end
+
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  # return array of all nodes in order
+  # when? if you need to print all the nodes of a tree in order 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list) if current_node.left
+    list << {key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, list) if current_node.right
+    
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  # when? if you need to save a tree data structure to disk, or send it across the network and maintain the structure
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list) if current_node.left
+    preorder_helper(current_node.right, list) if current_node.right
+    
+    return list
+  end
+  
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  # when? if you need to delete all the nodes in BST
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    postorder_helper(current_node.left, list) if current_node.left
+    postorder_helper(current_node.right, list) if current_node.right
+    list << {key: current_node.key, value: current_node.value}
+    
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  def height_helper(current_node, count, max)
+    return max if current_node.nil?
+
+    max = count if count > max
+    height_helper(current_node.left, count + 1, max)
+    height_helper(current_node.right, count + 1, max)
+  end
```

Comment:
This isn't working since it's only returning the right side of the traversal.

Consider

```ruby
def height_helper(current_node)
  return 0 if current_node.nil?

  left_height = height_helper(current_node.left)
  right_height = height_helper(current_node.right)

  return 1 + [left_height, right_height].max
end
```